### PR TITLE
Release LoRA usage for aborted waiting-queue requests

### LINKED ISSUE #29179: [Bug] LoRA request cancellation `/abort_request` can hang original `/generate` HTTP responses under concurrency
### Checklist

- [x] I searched related issues but found no solution.
- [x] The bug persists in the latest version.
- [x] Issues without environment info and a minimal reproducible demo are hard to resolve and may receive no feedback.
- [x] If this is not a bug report but a general question, please start a discussion at https://github.com/sgl-project/sglang/discussions. Otherwise, it will be closed.
- [x] Please use English. Otherwise, it will be closed.

### Describe the bug

When concurrent `/generate` requests use LoRA adapters, aborting them with `/abort_request` can leave the original `/generate` HTTP call blocked for at least 300 seconds even though:

- `/abort_request` returns HTTP 200,
- `/health` remains HTTP 200,
- `/v1/loads?include=lora,core,queues` reports `num_running_reqs=0` and `num_waiting_reqs=0`,
- a later normal LoRA request can still succeed in the abort-only repro.

The same abort pattern on base-model requests does not reproduce the response hang, so this appears specific to the LoRA request cancellation path. I reproduced this with raw `curl --max-time 300`, so it does not appear to be a Python HTTP client connection-pool issue.

### Reproduction

Please see the [repro.py](https://github.com/user-attachments/files/29294425/repro.py)

Minimal reproducing command used locally:

```bash
python repro.py \
  --disable-lora-overlap-loading \
  --disable-mutations \
  --rounds 30 \
  --concurrency 8 \
  --worker-generate-timeout-sec 10 \
  --request-timeout-sec 60 \
  --mem-fraction-static 0.40 \
  --context-length 4096 \
  --port 35213
```

The script:

1. Generates several local PEFT-style LoRA adapters.
2. Starts SGLang with LoRA enabled.
3. Preloads the adapters.
4. Sends concurrent long `/generate` requests, each with a `rid` and `lora_path`.
5. Sends `/abort_request` for each `rid` shortly after request submission.
6. Checks whether the original `/generate` call finishes within `--worker-generate-timeout-sec`.

## Observed Result

For LoRA requests, many original `/generate` calls were still blocked after their matching abort returned HTTP 200:
result: see [lora_overlap_abort_20260624_191411.json](https://github.com/user-attachments/files/29294521/lora_overlap_abort_20260624_191411.json)

```text
overlap=False, disable_mutations=True
rounds=30, concurrency=8
hung generate responses after abort: 14
final /health: 200
final queues: num_running_reqs=0, num_waiting_reqs=0
post-run base generate: 200
post-run LoRA generate: 200
```

With LoRA overlap loading enabled, the same abort-only pattern also reproduced:
result: see [lora_overlap_abort_20260624_191735.json](https://github.com/user-attachments/files/29294527/lora_overlap_abort_20260624_191735.json)

```text
overlap=True, disable_mutations=True
rounds=30, concurrency=8
hung generate responses after abort: 13
final /health: 200
post-run LoRA generate: 200
```

### Environment

Python: 3.11.14 (main, Oct 21 2025, 18:31:21) [GCC 11.2.0]
CUDA available: True
GPU 0,1,2,3: NVIDIA L40S
GPU 0,1,2,3 Compute Capability: 8.9
CUDA_HOME: /usr/local/cuda
PyTorch: 2.11.0+cu130
sglang: 0.5.13
sglang-kernel: 0.4.3
flashinfer_python: 0.6.12
flashinfer_cubin: 0.6.12
flashinfer_jit_cache: Module Not Found
triton: 3.6.0
transformers: 5.8.1
torchao: 0.17.0
numpy: 2.3.4
aiohttp: 3.13.2
fastapi: 0.121.1
huggingface_hub: 1.17.0
interegular: 0.3.3
modelscope: 1.31.0
orjson: 3.11.4
outlines: 0.1.11
packaging: 25.0
psutil: 7.1.3
pydantic: 2.12.4
python-multipart: 0.0.20
pyzmq: 27.1.0
uvicorn: 0.38.0
uvloop: 0.21.0
vllm: Module Not Found
xgrammar: 0.2.1
openai: 2.6.1
tiktoken: 0.12.0
anthropic: 0.72.0
litellm: Module Not Found
torchcodec: 0.11.1
NVIDIA Topology: 
        GPU0    GPU1    GPU2    GPU3    CPU Affinity    NUMA Affinity GPU NUMA ID
GPU0     X      PHB     NODE    NODE    0-127   0               N/A
GPU1    PHB      X      NODE    NODE    0-127   0               N/A
GPU2    NODE    NODE     X      PHB     0-127   0               N/A
GPU3    NODE    NODE    PHB      X      0-127   0               N/A

Legend:

  X    = Self
  SYS  = Connection traversing PCIe as well as the SMP interconnect between NUMA nodes (e.g., QPI/UPI)
  NODE = Connection traversing PCIe as well as the interconnect between PCIe Host Bridges within a NUMA node
  PHB  = Connection traversing PCIe as well as a PCIe Host Bridge (typically the CPU)
  PXB  = Connection traversing multiple PCIe bridges (without traversing the PCIe Host Bridge)
  PIX  = Connection traversing at most a single PCIe bridge
  NV#  = Connection traversing a bonded set of # NVLinks

ulimit soft: 524288

### PR DESCRIPTION
**Pull request by ronhuafeng**
## Motivation

Fixes #29179.

LoRA requests acquire a LoRA registry usage counter before dispatch. On the
normal completion path, `TokenizerManager._handle_batch_output()` releases that
usage when the request finishes.

The waiting-queue abort path is different: the scheduler echoes an `AbortReq`
back to tokenizer manager, and `TokenizerManager._handle_abort_req()` completes
the original `/generate` waiter directly. Before this PR, that path removed the
request state and notified the waiter, but did not release the LoRA registry
usage counter. A leaked counter can make later LRU unload block in
`wait_for_unload()`, which matches the reported symptom: `/abort_request`
returns 200, health stays OK, queues look empty, but later LoRA generation can
hang.

I reproduced the issue on current main with an H200 validation workload: 13 of
30 LoRA `/generate` calls stayed alive after their matching `/abort_request`
returned 200, and a post-run LoRA generation timed out. The base-only abort
control did not reproduce the hang.

## Modifications

- Release LoRA registry usage in `TokenizerManager._handle_abort_req()` when a
  scheduler-confirmed waiting-queue abort completes a LoRA request.
- Add unit coverage that:
  - a LoRA waiting-queue abort releases the acquired LoRA id exactly once;
  - a non-LoRA abort does not touch the LoRA registry.

## Accuracy Tests

This change does not alter model forward computation, sampling, logits, or
tokenization. It only releases LoRA usage bookkeeping after a scheduler-confirmed
abort.

H200 runtime validation with `Qwen/Qwen2.5-0.5B-Instruct` and generated LoRA
adapters:

- Baseline current main: LoRA abort workload reproduced the hang (`13/30`
  original `/generate` calls still alive after abort 200; post-run LoRA generate
  timed out).
- Patched, LoRA overlap disabled: `0/30` hung after abort, final `/health` 200,
  post-run base generate 200, post-run LoRA generate 200.
- Patched, base-only abort control: `0/12` hung after abort, final `/health`
  200, post-run base generate 200, post-run LoRA generate 200.
- Patched, LoRA overlap enabled: `0/30` hung after abort, final `/health` 200,
  post-run base generate 200, post-run LoRA generate 200.

The validation harness also observes an unrelated `/v1/loads?include=lora`
`slots_total=0` field warning after idle in both patched controls; the
user-visible hang signal is gone.

## Speed Tests and Profiling

No speed benchmark was run. The change only schedules existing LoRA registry
cleanup on a request abort path and should not affect normal generation speed.

## Checklist

- [x] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [x] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [x] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [x] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [x] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

Notes:

- Documentation is not changed because this fixes internal request cleanup.
- Accuracy and speed sections explain why model-output and throughput benchmarks
  are not applicable.

Validation commands:

```bash
git diff --check
python3 -m py_compile \
  python/sglang/srt/managers/tokenizer_manager.py \
  test/registered/unit/managers/test_tokenizer_manager_rid_cleanup.py
pre-commit run --files \
  python/sglang/srt/managers/tokenizer_manager.py \
  test/registered/unit/managers/test_tokenizer_manager_rid_cleanup.py
PYTHONPATH=python python3 -m pytest \
  test/registered/unit/managers/test_tokenizer_manager_rid_cleanup.py -q
```

---
### CI States

Latest PR Test (Base): :x: [Run #28346530955](https://github.com/sgl-project/sglang/actions/runs/28346530955)
Latest PR Test (Extra): :x: [Run #28346530866](https://github.com/sgl-project/sglang/actions/runs/28346530866)

### CODE REVIEW

File: python/sglang/srt/managers/tokenizer_manager.py [hunk 1]
```diff
@@ -2745,6 +2745,13 @@ def _handle_abort_req(self, recv_obj: AbortReq):
         }
         del self.rid_to_state[recv_obj.rid]
 
+        # Waiting-queue aborts do not go through _handle_batch_output, so release
+        # the LoRA usage acquired before dispatch once the scheduler confirms
+        # this request is no longer pending.
+        lora_id = getattr(state.obj, "lora_id", None)
+        if self.server_args.enable_lora and state.obj.lora_path and lora_id is not None:
+            asyncio.create_task(self.lora_registry.release(lora_id))
```
[the remark this answers is not in the document]
Comment (ronhuafeng) [reply python/sglang/srt/managers/tokenizer_manager.py:2754]: Good catch. Updated the current head for both parts:

- the waiting-queue abort release path now reads `lora_path` with `getattr`, matching the defensive `lora_id` lookup;
- `_discard_pending_req_states()` now releases acquired LoRA usage for still-pending states when dispatch fails after LoRA acquisition but before scheduler handoff.

I also added focused regression coverage for a missing `lora_path` attribute and for early dispatch failure after LoRA acquisition.

Local checks: `git diff --check` passed. I attempted the focused rid cleanup unittest, but this clean local Python environment is missing `msgspec`, so I could not execute it here.

**Comment by ronhuafeng** [comment]
I ran a baseline-vs-patched LoRA abort runtime check for this PR.

Setup:
- Hardware used for the active runtime: 1x NVIDIA H200.
- Image: `lmsysorg/sglang:latest@sha256:5027e95bf6ec536856b1b52a91d1f35ff5c564ab83e8a94758a169ff09bb8df3`.
- Model: `Qwen/Qwen2.5-0.5B-Instruct`.
- Base ref: `06fd2efedd7fab389c446d9cde17973e7650eb7f`.
- PR head tested: `6b911e68e19ebdea41ac7f1294a6971788442adf`.
- LoRA trigger: rank 128 adapters targeting Q/V projection, `--max-loaded-loras 2`, `--max-loras-per-batch 2`, `--disable-lora-overlap-loading`, mutations disabled.
- Workload: 30 abort rounds, concurrency 8, worker generate timeout 10s, request timeout 60s.

Results:
- Baseline failed as expected with 10 anomalies:
  - 9 abort cases returned HTTP 200 from abort but the corresponding generate request still hung after abort.
  - 1 post-LoRA generate request timed out after the abort workload.
- Patched passed with 0 anomalies.

This validates that request usage cleanup after LoRA aborts no longer leaves hung generation or a poisoned post-abort LoRA generation path in this focused runtime trigger.
